internal/cmd/auth.go

```diff
@@ -0,0 +1,117 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"go.flipt.io/flipt/internal/cleanup"
+	"go.flipt.io/flipt/internal/config"
+	"go.flipt.io/flipt/internal/gateway"
+	"go.flipt.io/flipt/internal/server/auth"
+	authtoken "go.flipt.io/flipt/internal/server/auth/method/token"
+	storageauth "go.flipt.io/flipt/internal/storage/auth"
+	storageoplock "go.flipt.io/flipt/internal/storage/oplock"
+	rpcauth "go.flipt.io/flipt/rpc/flipt/auth"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+func authenticationGRPC(
+	ctx context.Context,
+	logger *zap.Logger,
+	cfg config.AuthenticationConfig,
+	store storageauth.Store,
+	oplock storageoplock.Service,
+) (GRPCRegisterers, []grpc.UnaryServerInterceptor, func(context.Context) error, error) {
+	var (
+		register = GRPCRegisterers{
+			auth.NewServer(logger, store),
+		}
+		interceptors []grpc.UnaryServerInterceptor
+		shutdown     = func(context.Context) error {
+			return nil
+		}
+	)
+
+	// register auth method token service
+	if cfg.Methods.Token.Enabled {
+		// attempt to bootstrap authentication store
+		clientToken, err := storageauth.Bootstrap(ctx, store)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("configuring token authentication: %w", err)
+		}
+
+		if clientToken != "" {
+			logger.Info("access token created", zap.String("client_token", clientToken))
+		}
+
+		register.Add(authtoken.NewServer(logger, store))
+
+		logger.Debug("authentication method \"token\" server registered")
+	}
+
+	// only enable enforcement middleware if authentication required
+	if cfg.Required {
+		interceptors = append(interceptors, auth.UnaryInterceptor(
+			logger,
+			store,
+		))
+
+		logger.Info("authentication middleware enabled")
+	}
+
+	if cfg.ShouldRunCleanup() {
+		cleanupAuthService := cleanup.NewAuthenticationService(
+			logger,
+			oplock,
+			store,
+			cfg,
+		)
+		cleanupAuthService.Run(ctx)
+
+		shutdown = func(ctx context.Context) error {
+			logger.Info("shutting down authentication cleanup service...")
+
+			return cleanupAuthService.Shutdown(ctx)
+		}
+	}
+
+	return register, interceptors, shutdown, nil
+}
+
+func authenticationHTTPMount(
+	ctx context.Context,
+	cfg config.AuthenticationConfig,
+	r chi.Router,
+	conn *grpc.ClientConn,
+) error {
+	var (
+		muxOpts    []runtime.ServeMuxOption
+		middleware = func(next http.Handler) http.Handler {
+			return next
+		}
+	)
+
+	mux := gateway.NewGatewayServeMux(muxOpts...)
+
+	if err := rpcauth.RegisterAuthenticationServiceHandler(ctx, mux, conn); err != nil {
+		return fmt.Errorf("registering auth grpc gateway: %w", err)
+	}
+
+	if cfg.Methods.Token.Enabled {
+		if err := rpcauth.RegisterAuthenticationMethodTokenServiceHandler(ctx, mux, conn); err != nil {
+			return fmt.Errorf("registering auth grpc gateway: %w", err)
+		}
+	}
+
+	r.Group(func(r chi.Router) {
+		r.Use(middleware)
```

Comment:
Note to reviewers. I left this in place, even though no middleware is being registered yet.
In the OIDC PRs I register custom middleware to handle some of the more nuanced HTTP changes required to support OAuth in gateway.